### PR TITLE
t3c: mustache template DS parameter for remap.config lines

### PR DESCRIPTION
**Pull request by traeak**
This allows remap.config line to be reordered, modified, or plain hard coded set by specifying a DS parameter with a custom mustache template. (github.com/cbroglie/mustache)

The DS parameter is Config File: 'remap.config' and the Name one of:
- line_template.first
- line_template.inner
- line_template.last

The corresponding default values are as follows:

const DefaultFirstRemapConfigTemplateString = `map {{{Source}}} {{{Destination}}} {{{Strategy}}} {{{Dscp}}} {{{HeaderRewrite}}} {{{DropQstring}}} {{{Signing}}} {{{RegexRemap}}} {{{Cachekey}}} {{{RangeRequests}}} {{{Pacing}}} {{{RawText}}}`
const DefaultLastRemapConfigTemplateString = `map {{{Source}}} {{{Destination}}} {{{Strategy}}} {{{HeaderRewrite}}} {{{Cachekey}}} {{{RangeRequests}}} {{{RawText}}}`
const DefaultInnerRemapConfigTemplateString = DefaultLastRemapConfigTemplateString

Note this field may be used/abused to hard set the entire remap.config line for the given delivery service.

<hr/>

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Control Cache Config (`t3c`, formerly ORT)
- Traffic Portal

## PR submission checklist
- [x] This PR has tests 
- [x] This PR has documentation 
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

**Comment by ocket8888** [comment]
Why did you decide on this library? The one linked to on [the mustache website](https://mustache.github.io) is [cbroclie/mustache](https://github.com/cbroglie/mustache).

**Comment by traeak** [comment]
> Why did you decide on this library? The one linked to on [the mustache website](https://mustache.github.io) is [cbroclie/mustache](https://github.com/cbroglie/mustache).

It was one I had already used for some internal work.  I'll look at swapping out for this one.  This really is just playing around, I made it draft for others to look at.

**Review by zrhoffman — CHANGES_REQUESTED** [review]
Looks close! Is fixing the failing integration test  possible?

```go
--- FAIL: TestT3cGit (5.63s)
--- FAIL: TestT3cBadassAndSyncDs (11.31s)
--- FAIL: TestT3cBadassAndSyncDs/check_diff_of_remap.config_between_baseline_and_badass-generated (0.00s)
    t3c_mode_test.go:84: /opt/trafficserver/etc/trafficserver/remap.config and baseline-configs/remap.config differ:
```
```diff
diff --git a/baseline-configs/remap.config b/opt/trafficserver/etc/trafficserver/remap.config
index 6cee4fa..c0e5480 100644
--- a/baseline-configs/remap.config
+++ b/opt/trafficserver/etc/trafficserver/remap.config
@@ -1,5 +1,6 @@
-# do not edit - generated for atlanta-edge-03 by  () on fri nov 13 18:49:10 utc 2020
-map http://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/     http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-origin-ip-0' topology 'mso-topology'
-map https://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/     http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-origin-ip-0' topology 'mso-topology'
-map http://atlanta-edge-03.ds-top.test.cdn1.net/     http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-top' topology 'mso-topology'
-map https://atlanta-edge-03.ds-top.test.cdn1.net/     http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-top' topology 'mso-topology'
+# do not edit - generated for atlanta-edge-03 by t3c-generate 7.1.0..8a9a3388 from https://to_server:443 ips (172.18.0.3:443) on 2023-01-05t16:32:14.150696352z
+
+map http://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/    # ds 'ds-origin-ip-0' topology 'mso-topology'
+map https://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/    # ds 'ds-origin-ip-0' topology 'mso-topology'
+map http://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/    # ds 'ds-top' topology 'mso-topology'
+map https://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/    # ds 'ds-top' topology 'mso-topology'
```

File: lib/go-atscfg/remapdotconfig.go [hunk 1]
```diff
@@ -29,16 +30,66 @@ import (
 
 	"github.com/apache/trafficcontrol/lib/go-tc"
 	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/cbroglie/mustache"
 )
 
-const CacheKeyParameterConfigFile = "cachekey.config"
+const CachekeyParameterConfigFile = `cachekey.config`
```
[the remark this answers is not in the document]
Comment (traeak) [reply lib/go-atscfg/remapdotconfig.go]: removed

File: lib/go-atscfg/remapdotconfig.go [hunk 2]
```diff
@@ -29,16 +30,66 @@ import (
 
 	"github.com/apache/trafficcontrol/lib/go-tc"
 	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/cbroglie/mustache"
```
[the remark this answers is not in the document]
Comment (traeak) [reply lib/go-atscfg/remapdotconfig.go:34]: fixed up.

**Review by zrhoffman — CHANGES_REQUESTED** [review]
Very close, there's just
* 2 spaces between `@pparam=hdr_rw_first_ds-origin-ip-0.config` and `@plugin=cachekey.so` in the generated `remap.config`
* 2 spaces between `@pparam=--capture-prefix-uri=/^([^?]*)/$1/` and `#` in the generated `remap.config`

If you want to just add those spaces to the reference, the test would pass and we could call it good
```go
--- FAIL: TestT3cBadassAndSyncDs/check_diff_of_remap.config_between_baseline_and_badass-generated (0.00s)
    t3c_mode_test.go:84: /opt/trafficserver/etc/trafficserver/remap.config and baseline-configs/remap.config differ:
```
```diff
diff --git a/baseline-configs/remap.config b/opt/trafficserver/etc/trafficserver/remap.config
index 0352abf..6df6491 100644
--- a/baseline-configs/remap.config
+++ b/opt/trafficserver/etc/trafficserver/remap.config
@@ -1,5 +1,6 @@
-# do not edit - generated for atlanta-edge-03 by  () on fri nov 13 18:49:10 utc 2020
-map http://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-origin-ip-0' topology 'mso-topology'
-map https://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-origin-ip-0' topology 'mso-topology'
-map http://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-top' topology 'mso-topology'
-map https://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/ # ds 'ds-top' topology 'mso-topology'
+# do not edit - generated for atlanta-edge-03 by t3c-generate 7.1.0..c3af4984 from https://to_server:443 ips (172.18.0.3:443) on 2023-01-05t21:01:49.85368428z
+
+map http://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config  @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/  # ds 'ds-origin-ip-0' topology 'mso-topology'
+map https://atlanta-edge-03.ds-origin-ip-0.test.cdn1.net/ http://192.0.2.1/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-origin-ip-0.config  @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/  # ds 'ds-origin-ip-0' topology 'mso-topology'
+map http://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config  @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/  # ds 'ds-top' topology 'mso-topology'
+map https://atlanta-edge-03.ds-top.test.cdn1.net/ http://origin.topology.example.net/ @plugin=header_rewrite.so @pparam=dscp/set_dscp_40.config @plugin=header_rewrite.so @pparam=hdr_rw_first_ds-top.config  @plugin=cachekey.so @pparam=--separator= @pparam=--remove-all-params=true @pparam=--remove-path=true @pparam=--capture-prefix-uri=/^([^?]*)/$1/  # ds 'ds-top' topology 'mso-topology'
```

**Review by zrhoffman — APPROVED** [review]
Looks good, thanks!

**Pull request merged**